### ISSUE TITLE
Chakra JS data structure documentation

### ISSUE THREAD

**Issue by nojvek** [Microsoft/ChakraCore#306]
It would be great if there was some documentation on how chakra implements the JS datastructures such as arrays and objects and what goes on in the JIT world.

**Comment by ianwjhalliday** [comment]
Agreed!  A blog would be a great forum for this.  I'm not committing to anything here, since it would definitely be pressure on resources, but I think this is something we should seriously consider.

**Comment by nojvek** [comment]
I understand. Would it work if I interviewed the team and took some notes? May be turn it into a little gitbook or something. The value of this is when someone is working on performance sensitive javascript, its important to understand what is going on under the hood. e.g why is new Array(1000) a lot faster than a 100 array.push calls.

**Comment by digitalinfinity** [comment]
We have some old blog posts and talks that are still somewhat relevant- are you thinking something along these lines?
https://blogs.msdn.microsoft.com/ie/2012/06/13/advances-in-javascript-performance-in-ie10-and-windows-8/
https://www.youtube.com/watch?v=OqnMmoF2GqM

**Comment by nojvek** [comment]
Hi Hitesh,

That's still somewhat a high level overview. I was more interested in how a
JavaScript object is stored in memory? How do you ensure fast property
access? When do bailouts happen? What happens when more properties are
added to an object. How to make large hash maps efficient? . How does array
allocations work on .push? How do you ensure functions get inlined? How do
you profile chakra e.t.c

An example is I am writing a simple virtual Dom diffing library to achieve
fast speeds. How do I optimize my js for fast perf and efficient garbage
collection. To do that understanding what goes under the hood in js runtime
would be very valuable for any developer.

**Comment by sankha93** [comment]
@abchatra has written some posts regarding this which provide some insight into JS engine internal data structures. You can find them at http://abchatra.github.io/

**Comment by abchatra** [comment]
Let me know if you have any questions. I will be glad to help. Also if you are interested in particular topic do let me know.

**Comment by nojvek** [comment]
Thanks Abhijit,

Thanks for the blogs. There's some great stuff there.  Looking forward to
inline cache article.

This is the kind of technical stuff that I think is very interesting.

I'm personally interested to learn how arrays and object maps work. How is
their runtime and memory performance of different operations.

How functions get stored and called. When do they get inlined? How scope
inheritance is calculated?

Basically how can I write my JavaScript so it performs at the speed close
to something written in C++.

**Comment by am06** [comment]
+1. I'd like to read more stuff about this too. That blog was great, too bad it had only 2 posts though.
